### ISSUE TITLE
LUA Telemetry for Crossfire

### ISSUE THREAD

**Issue by advancetom** [teckel12/LuaTelemetry#36]
Hey there i found this treat but cant working with my crossfire micro V2.
I saw my sensors but the lua says no telemetry!!!
It is only working with SPort not Crossfire yet ?

**Comment by teckel12** [comment]
I believed the docs were clear that this script is only for SmartPort telemetry. It's not that it's not yet Crossfire compatible, it's that Crossfire isn't SmartPort telemetry and therefore not compatible.

It would seem that someone with Crossfire hardware would need to rewrite the crossfire telemetry to be compatible with SmartPort telemetry and then it would be compatible by default with this script. But, that would also probably break any existing Crossfire telemetry scripts and transmitter model setup.

I don't have Crossfire hardware to know if compatibility would be possible.

Let me know if you feel Crossfire is compatible with SmartPort and it should work.

**Comment by teckel12** [comment]
I just read this issue for OpenTX:

https://github.com/opentx/opentx/issues/4922

It seems that there may be issues with Crossfire telemetry if using the Taranis Q X7.  No telemetry would be more at the transmitter level than the lua script level.  Do normal telemetry screens work for you?

**Comment by advancetom** [comment]
Yes my normal Telemetry Screen Works Fine

**Comment by teckel12** [comment]
@advancetom Isn't Crossfire a totally different telemetry from SmartPort?  You obviously use a different transmitter module and receiver.  For telemetry to be considered good with the LuaTelemetry script, the telemetry sensor `RSSI` must be returning a positive value.  What does RSSI show?  Is it labeled `RSSI` (case sensitive)?

- What telemetry sensor names do you get on your transmitter?  
- What telemetry values do you get for telemetry sensors `Tmp1` and `Tmp2`?

**Comment by advancetom** [comment]
Downlink
TRSS | Signalstärke Downlink
TQly | Empfangsqualität Sender
TSNR | Downlink SNR
GPS | GPS Koordinaten
Alt | Höhe über NN
Sats | Anzahl Satelliten
RxBt | Akkuspannung
Curr | Aktueller Strom
Capa | Akkukapazität
Ptch | Pitch Winkel
Roll | Roll Winkel
Yaw | Yaw Winkel

Uplink

Uplink | zum Empfänger
1RSS | Signalstärke Antenne 1
2RSS | Signalstärke Antenne 2 (nur bei Diversity Empfänger)
RQly | Empfangsqualität Empfänger
RSNR | Uplink SNR
RFMD | 0= 4Hz; 1= 50Hz, 2= 150Hz Aktualisierungsrate
TPWR | Sendeleistung

This are the values from crsf protokoll (Crossfire)
Tmp1 and Tmp2 i dont have it

**Comment by teckel12** [comment]
Yeah, those are totally different and there's some critical ones missing. Crossfire telemetry in INAV would need to be updated to work with the Lua Telemetry script.

I'll look at the INAV code and see if additional sensors are even available.

**Comment by teckel12** [comment]
@advancetom I just looked at the Crossfire telemetry code. It's sending Crossfire-specific telemetry packages that are then deconstructed by the Crossfire module in your transmitter to show as telemetry values. Basically, it's not possible without new firmware in the transmitter module to accommodate additional flight information from INAV. Not impossible, but highly unlikely.

**Comment by KnuckleUpFPV** [comment]
https://github.com/tgreer/inav-telemscript


This is a crossfire inav script if anyone here is interested. Qx7 version is coming.

**Comment by teckel12** [comment]
@KnuckleUpFPV Actually, looking at that source code, it seems that maybe Crossfire support could be forced into Lua Telemetry.  As I don't have Crossfire I've never seen what the actual transmitter-side telemetry values you get with Crossfire.  This source code gives a lot of insight.

The main potential problem is the flight mode.  It's sent as a string, and I'm not sure of all the possibilities and if they match with the logic in Lua Telemetry enough to make it work.  But, if I could get a telemetry log and Taranis config from someone who's flying Crossfire I could at least setup a simulation and then see if shoehorning Crossfire into Lua Telemetry is possible.

Below are the instructions for exporting radio and model settings that are needed for a simulation:

1. Connect transmitter to PC using standard method (trim in, power, connect USB) see: https://youtu.be/GAETmZsKqwY?t=2m13s
1. Run OpenTX Companion
1. From 'Read/Write' menu, select 'Read Models and Settings From Radio'
1. From 'File' menu, select 'Save As...' and save radio models and settings to OTX file
1. ZIP OTX file and attach it to a new comment in this issue
1. Make sure you specify the model number/name from your transmitter

I'll also need some telemetry log files.  If you're not currently creating telemetry log files, each of your telemetry sensors need to have the `Logs` box checked (I believe that's the default).  Then, you just need to setup a special function to record SD Logs.  I set this to when my arm switch is in the arm position `SF↓` (use the arm switch for your model), then set the function to `SD Logs` and the time to `0.1s` or `0.2s`.  As the function is set to record when the arm switch is enabled, it automatically saves log files to your SD card's `LOG` folder for each flight.

Include a few of the telemetry log files along with the ZIPped OTX file created above will allow me to simulate your entire flight in OpenTX Companion.

**Comment by teckel12** [comment]
### Resources
https://github.com/tgreer/inav-telemscript/blob/master/incrsf.lua#L590
https://github.com/iNavFlight/inav/blob/master/src/main/telemetry/crsf.c#L229
https://github.com/iNavFlight/inav/blob/master/src/main/io/osd.c#L1471

It looks like the flight modes are VERY limited:

```
ACRO
ANGL
HOR
!FS
```

There's been some interest for Crossfire telemetry flight modes to match the OSD flight mode strings.  That seems logical but I've yet to see any work done on this.

I could add support in INAV for enhanced flight modes (and matching the OSD flight modes).  I'd just need someone with Crossfire to test my changes before submitting a PR to INAV.  I'd be doing this with INAV 2.0 so the only requirement is that you're currently running INAV 2.0.  Just let me know what target you use and I can build an enhanced flight mode HEX file for you to test.

I'll look at what essential flight modes would be required for Lua Telemetry to work.  It doesn't seem right now that Crossfire even gives an indication when you're armed, so that at least would be a must.  

These enhancements should help everyone with Crossfire, not just those running Lua Telemetry.

**Comment by teckel12** [comment]
@tgreer Interested in helping out?  Seems you could even help with the coding too, which would be great as I've had only one pull request for Lua Telemetry that wasn't from me.

**Comment by KnuckleUpFPV** [comment]
I have months worth of telemetry logs with gps and barometer info from wings and quads. If i can figure out how to attach the file here im happy to do so. If not i can email it to you.

**Comment by teckel12** [comment]
@KnuckleUpFPV You should be able to follow these instructions to save your transmitter's OTX file with all the model/config information:

1. Connect transmitter to PC using standard method (trim in, power, connect USB) see: https://youtu.be/GAETmZsKqwY?t=2m13s
1. Run OpenTX Companion
1. From 'Read/Write' menu, select 'Read Models and Settings From Radio'
1. From 'File' menu, select 'Save As...' and save radio models and settings to OTX file

This will be a single OTX file.  Then, you can just copy over the files from your `LOG` folder from your transmitter's SD card to your computer.

1. From a folder on your computer with all the log files and OTX file
1. Select all the log files and the OTX file
1. Right click 
1. Select `Send to`
1. Select `Compressed (zipped) folder`

Then just drag/drop the created ZIP file to a reply here and it will auto-upload and attach it.

**Comment by KnuckleUpFPV** [comment]
I know how to get the files. Getting it here was the issue. I can take care of that tomorrow after work. Didnt know you could drag and drop. Im only ever on github on my phone. Ill get you what you need tomorrow.

**Comment by tgreer** [comment]
I wrote my own instead.

**Comment by teckel12** [comment]
@KnuckleUpFPV Sounds good! What target is your FC so I can get you a custom build with additional flight modes?

Also, do you have 2 RSSI sensors? Normal for CR? Would the average or lowest one be the value displayed?

**Comment by teckel12** [comment]
@tgreer looks nice. It looks familiar, is it a fork or another project modified to Crossfire telemetry sensors or did you combine a couple different projects?

**Comment by KnuckleUpFPV** [comment]
Omnibus f4 pro is the target i have the most fcs of. We have two rssi readings, and a link quality. Link quality is far more accurate than rssi. I will be home in a little bit to get the info to you. Another hour and im off of work.

**Comment by teckel12** [comment]
@KnuckleUpFPV Oh, are you running INAV 2.0 final?

**Comment by KnuckleUpFPV** [comment]
Im running 1.8, and 1.9.1. I wont have any flights on 2.0 till this weekend most likely. Just put my flite test goblin and my s800 on 2.0. Ill let you know what firmware version i send telemetry for tonight.

**Comment by teckel12** [comment]
@KnuckleUpFPV No hurry, I'm still sorting out the code and what Lua Telemetry needs at a minimum.

**Comment by teckel12** [comment]
Justification for enhancement to crsf telemetry: https://github.com/iNavFlight/inav/issues/3781

**Comment by KnuckleUpFPV** [comment]
[inav lua crossfire.zip](https://github.com/iNavFlight/LuaTelemetry/files/2448511/inav.lua.crossfire.zip)

**Comment by KnuckleUpFPV** [comment]
the s800 is an omnibus f4 pro v2
the sonic model ar is an matekf405 aio

if you need anything else let me know. 
The rssi or link quality can be selected and fed over channel 12. most use lq.

**Comment by tgreer** [comment]
Combination of others tweaked to work with inav and crossfire.

**Comment by KnuckleUpFPV** [comment]
I messed up.  I included two sonic model ar files. So its only matek f405 aio. Noisy gyro version. Ill attach an s800 tonight.

**Comment by KnuckleUpFPV** [comment]
So do i just flash this to my fc and then try your lua?

**Comment by teckel12** [comment]
@KnuckleUpFPV Flash your FC and when you fly monitor your flight mode telemetry sensor. Previously it would just show `ACRO`, `HOR`, `ANGL`. Now it will show `RDY` (ready), `NRDY` (not ready), and many different flight modes, for example `HOLD`, `RTH`, etc. 

Lua Telemetry won't *yet* work, as I'll need some telemetry logs with the  flight modes from this firmware to shoehorn in the Crossfire telemetry information.

The telemetry for Crossfire is quit different than SBUS. Lua Telemetry depends on this telemetry info to know if (for example) your ready to arm or not. Previously, Crossfire didn't even report if the craft was ready to arm, not ready, or armed. That's key information that Lua Telemetry requires to function. With this firmware, Lua Telemetry will be modified to look at the Crossfire telemetry and know the flight status and mode so it can operate.

With a few telemetry logs from your transmitter with this firmware, the process of making Lua Telemetry compatible with Crossfire is fairly trivial.

First, let me know if the firmware show additional telemetry flight modes, then supply some additional telemetry logs from your transmitter. I can then submit a pull request to INAV to include these changes in the next release. When I get the changes done in Lua Telemetry, I'll give instructions on how to install and test.

Let me know if there's any issues.

**Comment by KnuckleUpFPV** [comment]
Will do. Ill be flying this wing tomorrow so i will get it flashed and ready to go tonight. Should have the logs for you tomorrow evening.

**Comment by teckel12** [comment]
BTW, I'm not happy with the `RDY` and `NRDY` flight statuses.  I just put them in there as a temporary placeholders.  Now that I've thought about it a bit, ready to arm should probably be `OK`?  Still not sure about the `NRDY` (not ready) flight status.  I was thinking maybe `WAIT`?

How it should work is when you first power up your model it should be in the not ready flight status, so `WAIT` kind of makes sense.  Once it's ready to arm it would show `OK`, which also makes sense.  Once armed, it would show your flight status.  So the progression would be something like `WAIT`, `OK`, `ANGL` or whatever.

The only rub here is that `WAIT` could be forever (for example, if something isn't calibrated, the GPS isn't working or whatever the fault is).  So is `WAIT` still a good note or would some other 4 character note be better?

None of this really matters to Lua Telemetry, as it will take this flight information and then do it's own thing, displaying what it normally does and playing the correct voice file.  So this makes no difference to Lua Telemetry, but it could for people using Crossfire without Lua Telemetry.

**Comment by KnuckleUpFPV** [comment]
I think RTA for ready to arm. And wait would be fine.

**Comment by KnuckleUpFPV** [comment]
When betaflight isnt ready it tells you why. Like if you dont have enough sats it says gps.

**Comment by KnuckleUpFPV** [comment]
I think RTA for ready to arm. And wait would be fine.

**Comment by teckel12** [comment]
@KnuckleUpFPV Use this one, I discovered there was an issue with the previous build (don't have Crossfire so I couldn't test it first, sorry):

[inav_2.0.0_MATEKF405SE.hex.zip](https://github.com/iNavFlight/LuaTelemetry/files/2450974/inav_2.0.0_MATEKF405SE.hex.zip)

**Comment by teckel12** [comment]
@KnuckleUpFPV I'll look at the Betaflight source to see what it uses.  Best to use the same as much as possible.

**Comment by KnuckleUpFPV** [comment]
Ok

**Comment by KnuckleUpFPV** [comment]
Im positive it says gps if you are under 5 satellites and try to arm. It pops up right on the osd.  Crossfire only gives coordinates and number of satellites. I dont know what you mean by betaflight only adds air as flight mode. Its has stab horz and acro as flight modes. I have a logic setup on my transmitter. If Im waiting for sats with goggles off my transmitter says gps lock after it sees 5 or more satellites. The betaflight lua doesn't show anything but settings. I havent ever seen a betaflight lua like the inav and ardu lua.

**Comment by teckel12** [comment]
@KnuckleUpFPV The OSD has nothing to do with telemetry. And it's also not specific to Crossfire (or anything else). INAV will also show this information on OSD, and will give thr reason why it can't arm. 

Betaflight currently only has one ADDITIONAL flight mode over INAV, which is AIR. But with the firmware I posted, AIR is added along with many others.

I was under the impression that you were saying that Betaflight was giving you some information over Crossfire that INAV wasn't. But, that's not the case (other than AIR, which won't be the case after my pull request). After my pull request, INAV will give more information than betaflight for Crossfire telemetry.

**Comment by KnuckleUpFPV** [comment]
Yeah man sorry about that. Wasnt clear enough.

**Comment by teckel12** [comment]
@KnuckleUpFPV I've kinda went back and forth about this... But I think it makes sense to report if there's no GPS lock but no other reasons for not being able to arm.

The reasoning is that there are SO many reasons why you can't arm, it's just too difficult to do with only 4 characters and not very useful as almost everyone runs FPV so the OSD will tell you the reason in a whole line of text.

But for almost every flight, you first wait for a GPS lock, so it would be useful to know that the GPS is what you're waiting for. So I'm thinking when not non-armed there would be three conditions:

1) Waiting for GPS, flight mode `WAIT`? 
2) Ready to arm, flight mode `OK`?
3) Some other reason it can't arm, flight mode `!ARM` or `!ERR`?

I think this will be good enough for the majority of flights.

Logic sound? Missing something? Thoughts?

**Comment by KnuckleUpFPV** [comment]
I think that sounds great. Had some issues with the wing today. Drove an hour to not get it in the air. Was able to get my other wing with the omnibus in the air. If you have a hex for the omnibus f4 pro. It might be the better option right now. I can just go to the school tomorrow and sling it in the air.

**Comment by ZinhoRacing** [comment]
Just another point of view :)
The Flight mode is not limited to 4 characters with crossfire protocol.
Here is what i did (removing the details of unable to arm reasons)

`  // use same logic as OSD, so telemetry displays same flight text as OSD when armed
    const char *flightMode = "Ready to arm";
    if (ARMING_FLAG(ARMED)) {
				if (FLIGHT_MODE(FAILSAFE_MODE)) {
					flightMode = "!FAILSAFE!";
				} else if (FLIGHT_MODE(MANUAL_MODE)) {
					flightMode = "Manual";
				} else if (FLIGHT_MODE(NAV_RTH_MODE)) {
					flightMode = "Return Home";
				} else if (FLIGHT_MODE(NAV_POSHOLD_MODE)) {
					if (FLIGHT_MODE(NAV_ALTHOLD_MODE)) {
						// 3D HOLD
						flightMode = "3D Hold";
					} else {
						flightMode = "Pos Hold";
					}
				} else if (FLIGHT_MODE(NAV_CRUISE_MODE) && FLIGHT_MODE(NAV_ALTHOLD_MODE)) {
					flightMode = "3D Cruise";
				} else if (FLIGHT_MODE(NAV_CRUISE_MODE)) {
					flightMode = "Cruise";
				} else if (FLIGHT_MODE(NAV_ALTHOLD_MODE) && navigationRequiresAngleMode()) {
					// If navigationRequiresAngleMode() returns false when ALTHOLD is active,
					// it means it can be combined with ANGLE, HORIZON, ACRO, etc...
					// and its display is handled by OSD_MESSAGES rather than OSD_FLYMODE.
					flightMode = "AltitudeHold";
				} else if (FLIGHT_MODE(NAV_WP_MODE)) {
					flightMode = "Nav WayPoint";
				} else if (FLIGHT_MODE(ANGLE_MODE)) {
					flightMode = "ANGLE";
				} else if (FLIGHT_MODE(HORIZON_MODE)) {
					flightMode = "HORIZON";
				}else if (isAirmodeActive()) {
					flightMode = "AIR";
				}else {
					flightMode = "ACRO";
				}		
	} else
        {flightMode = "Unable to Arm"; }
`

I liked your idea associated to WAIT (for GPS fix) and !ERR for Error and not able to Arm.

**Comment by teckel12** [comment]
@ZinhoRacing While Crossfire may not be limited to 4 characters, that's what the OSD is limited to for flight modes. Also, the more characters, the slower your telemetry will be. So sending many characters will make all of your telemetry slow down for everything else, which may not be ideal. Finally, reasons for not arming are not flight modes even in OSD, but are error messages instead. 

Honestly, I'm not sure if reasons for not arming are really that useful or should be sent as telemetry. No other telemetry sends this information. Also, the info is already available on the OSD. Finally, they hardly ever come up, some really never. 

Basically, it seems like an unnessasary duplication of code from OSD. If it's included at all, it probably should be combined code with osd.c instead of being duplicated in both.

So, focusing on just the flight modes, I'd like to get that working. It will not only assist all Crossfire pilots, but also allow for lua scripts like Lua Telemetry to function (hence my involvement).

I think there should be a conversation about how to combine the error messages on the OSD, Crossfire, and maybe many other places (like OLED display for example). That would be a different pull request.

**Comment by ZinhoRacing** [comment]
Totaly agree with you @teckel12, I think we have the same view on that point.
Limiting to 4 characters will make telem speed stay in good health. That was my first idea when i thought of codes for flight modes (issu posted on Inav).
We only need to agree to a standardized list of flight modes to be able to make some scripts work (Lua Telem of course and I'm using one that I modifyed from yaapu for Horus).
About having some error messages available on telem is for me a nice to have, because some times I go to fly without googles or ground station Monitor. So if I'm able to see on my radio screen over a script a reason why I can't arm it would be very helpful. And again, i Agree with you it should be a different pull request.

**Comment by teckel12** [comment]
@ZinhoRacing @KnuckleUpFPV Here's the latest build:

[inav_2.0.0_MATEKF405SE.hex.zip](https://github.com/iNavFlight/LuaTelemetry/files/2455876/inav_2.0.0_MATEKF405SE.hex.zip)
[inav_2.0.0_OMNIBUSF4PRO.hex.zip](https://github.com/iNavFlight/LuaTelemetry/files/2455877/inav_2.0.0_OMNIBUSF4PRO.hex.zip)

And here's the changes to `crsf.c`:
```
    const char *flightMode = "OK";
    if (ARMING_FLAG(ARMED)) {
        if (isAirmodeActive()) {
            flightMode = "AIR";
        } else {
            flightMode = "ACRO";
        }
        if (FLIGHT_MODE(FAILSAFE_MODE)) {
            flightMode = "!FS!";
        } else if (FLIGHT_MODE(MANUAL_MODE)) {
            flightMode = "MANU";
        } else if (FLIGHT_MODE(NAV_RTH_MODE)) {
            flightMode = "RTH";
        } else if (FLIGHT_MODE(NAV_POSHOLD_MODE)) {
            flightMode = "HOLD";
        } else if (FLIGHT_MODE(NAV_CRUISE_MODE) && FLIGHT_MODE(NAV_ALTHOLD_MODE)) {
            flightMode = "3CRS";
        } else if (FLIGHT_MODE(NAV_CRUISE_MODE)) {
            flightMode = "CRS";
        } else if (FLIGHT_MODE(NAV_ALTHOLD_MODE) && navigationRequiresAngleMode()) {
            flightMode = "AH";
        } else if (FLIGHT_MODE(NAV_WP_MODE)) {
            flightMode = "WP";
        } else if (FLIGHT_MODE(ANGLE_MODE)) {
            flightMode = "ANGL";
        } else if (FLIGHT_MODE(HORIZON_MODE)) {
            flightMode = "HOR";
        }
#ifdef USE_GPS
    } else if (feature(FEATURE_GPS) && (!STATE(GPS_FIX) || !STATE(GPS_FIX_HOME))) {
        flightMode = "WAIT"; // Waiting for GPS lock
#endif
    } else if (isArmingDisabled()) {
        flightMode = "!ERR";
    }
```

**Comment by teckel12** [comment]
To do:
* [x] Identify telemetry sensors
* [x] Make changes to INAV so arming and GPS status are sent along with expanded flight modes
* [x] Create detection and remapping/conversion of Crossfire telemetry to standard SmartPort telemetry
* [x] Testing

**Comment by ZinhoRacing** [comment]
What is you idea on first point ? On my side for Horus script I used MSP over crossfire CRSF protocol from code that i merged from betaflight to have the arming status, GPS status (fix type, alt, speed) and distance / direction to home available. I thought at beginning to add it to flight mode too, but had no real idea how to do it. I wish I could help.

**Comment by teckel12** [comment]
@KnuckleUpFPV I fear getting Crossfire to work with Lua Telemetry will be a lot more challenging that I had thought.  Doing any type of development in the OpenTX Companion will be very difficult to impossible as the sensor names and values are all different or totally missing in Crossfire making it impossible to setup a development environment to write code against.  And writing code in a vacuum is a real challenge.

Looking at Crossfire telemetry logs, the following seem true and I need confirmation:
* RxBt = battery voltage?  This is different than SBUS, which VBAT is the battery voltage and RxBt is the voltage going to the receiver (more like 4.5 - 5 volts).
* GSpd and Hdg missing?  I'm not getting these telemetry sensors in the logs, do they work?
* Yaw is the same as Hdg?
* No distance from home info?
* No GPS altitude info?
* No vertical speed?

**Comment by teckel12** [comment]
@ZinhoRacing  I don't follow how you're getting the arming status, GPS status, fix type, alt speed, distance, etc.  I don't see any of those telemetry sensors on the radio.  Not sure what your doing to get that data, or how you would display it.  Maybe it's something Horus specific which I don't have the faintest idea how that would work as I don't have that transmitter nor does Lua Telemetry work with it because it's totally different.

It sounds like you're saying that in addition to the normal Crossfire telemetry you're also sending other information over the Crossfire telemetry?  Would this be something that all Crossfire users do?  Is it something that works with INAV and Taranis transmitters?  If it's something specific to Betaflight or Horus or something where you need to do something custom, it may require a developer (with this specific hardware) to write the code, which wouldn't be me as I don't understand a lick of how you're getting this data on your transmitter.

**Comment by KnuckleUpFPV** [comment]
He is using mavlink over bluetooth. The crossfire sends all that information over the bluetooth link as it receives it from the gps. Gspd and heading are reported to the taranis in the telemetry. Rxbat is battery voltage at the model. Sorry haven't had a chance to get your firmware in the air yet. Weekend ended up getting busy.

**Comment by ZinhoRacing** [comment]
@teckel12 @KnuckleUpFPV 
Ok, here is the full explanation. (not using bluethoot or Mavlink)
I came across the same missing values when i started to think of using a widget for Horus (I wanted to use LuaTelem but Horus is not supported yet, hope we will do it :) ).
My configuration is Inav2.0.0, using crossfire (crsf protocol) as radio link. On the radio, i'm using as well crossfire (OpenTx2.2.2). So I have a full Crsf protocol radio link from Inav to OpenTx. As you spotted, OpenTx or Inav are not sending/decoding all the values needed for LuaTelem (as for the widget i worked on, same values are missing). So i tryed to find a solution because changing OpenTx is very difficult. I find out that in Betaflight, They have been able to get MSP protocol to passthrough CRSF radio link (MSP protocol is used for Configurator and some light scripts to change calibration like PID). I then merged all the necessary code in Inav (custom made). It do work. I'll post part of the Lua code (telem function) to explain how i did tonight (I'm at work and don't have the source here with me.)

About crsf native telem here is what I understood analysing code in Inav and OpenTx and testing.

- **RxBt =** battery voltage, this is the pack voltage as read by the FC (source Inav code and tested OK)
- **GSpd and Hdg missing?** not working, I had to use MSP ( message MSP  106, MSP_RAW_GPS, getting GPS status(fix), GPS hspeed that is horizontal speed, GPS Hdop, GPS Alt)
- **Yaw is the same as Hdg?** yes, it is the heading (source from Inav code and tested OK) it is in radian 
Here is part of my Lua code for Attitude
``` status.roll = math.deg(getValue("Roll"))
	status.pitch = -math.deg(getValue("Ptch"))
	status.yaw = math.deg(getValue("Yaw"))
	if (status.yaw < 0 ) then
		status.yaw = status.yaw + 360
	end
```
- **No distance from home info?** I get it over MSP (message MSP 107, MSP_COMP_GPS, getting Home Dist and Home Angle)
- **No GPS altitude info?** I get it over MSP 106, Note that Home Altitude is Alt sensor in native CRSF telem.
- **No vertical speed?** I get it over MSP (MSP 109 MSP_ALTITUDE, this is the value used for OSD in Inav).
- **About Armed status**, I had to use another MSP message MSP 150 MSP_STATUS_EX, getting INAV ArmingFlags where bit 2 is the bit to know if it is Armed (in my code statusArmed). 
``` elseif (cmd == 150) then
     if #(rx_buf) > 0 then
         status.packSensorStatus = readrxbufferMSP (rx_buf, 5, 6, 1)
         status.armingFlags = readrxbufferMSP (rx_buf, 14, 15, 1)
         status.statusArmed = bit32.extract (status.armingFlags, 2, 1) 
     end
 end
```
So I have all the informations needed to make it work. Of course it still need improvement as I didn't look after rates and so, but It is a start.

When you will see the Lua code of the my Horus Widget, it will all be clear I hope :)

Summary :  To make it work, I changed Inav code (adding MSP capability over CRSF telem protocol, code from betaflight) +  coded CRSF + MSP telem decoding in the Horus Widget (Yaapu Widget).

**Comment by jorgermp** [comment]
> https://github.com/tgreer/inav-telemscript
> 
> This is a crossfire inav script if anyone here is interested. Qx7 version is coming.
@KnuckleUpFPV I have a qx7, if you send me the file I can test it.

**Comment by teckel12** [comment]
@jorgermp After INAV merges my changes, Lua Telemetry will work with Crossfire on the X9D, X7 and Horus transmitters.

The issue was never with Lua Telemetry, it was that Crossfire telemetry provided rudimentary flight information which doesn't give enough info for Lua Telemetry to function. Like actually knowing when the craft is armed as an example. This changes once INAV is updated in the next release.

**Comment by KnuckleUpFPV** [comment]
@teckel12  if I can get a matekf405 hex I can test this out. The matekf405se board I have was destroyed. Havent had much time to fly. Actually got out the other day and lost a wing for a whole week. Found it pierced on a tree limb.

**Comment by KnuckleUpFPV** [comment]
@teckel12  the script looks amazing by the way. Took me a second to figure out there were more screens to go through. Absolutely amazing job.

**Comment by KnuckleUpFPV** [comment]
i tested the omnibus hex posted here and it doesnt work unfortunately.

**Comment by teckel12** [comment]
@KnuckleUpFPV The HEX posed here is for testing the additional flight data information added to INAV. Basically, testing that the new flight modes are transmitted and displayed on your Taranis transmitter. That needs to be incorporated into INAV before a new release of Lua Telemetry that would understand what the new flight modes translate to. There's other things missing as well from Crossfire, like craft distance. Lua Telemetry calculates this based on launch GPS location.

Basically, the HEX is step 1, which we're not fully through as it's not yet been incorporated into INAV (should be in 2.1).

Whe you say it doesn't work, I assume you mean Lua Telemetry, which would be partly true. Look for the flight modes from telemetry sensors, that's what should be tested.

**Comment by KnuckleUpFPV** [comment]
It doesnt show anything in the lua. There is no. Flight mode or anything

**Comment by teckel12** [comment]
@KnuckleUpFPV Correct, that's not what the HEX is for.  The HEX is to test the telemetry sensors (something I thought you already did) not having anything to do with Lua at all.  The important telemetry sensor to test (and the only one changed in the HEX) is the flight mode.

Lua Telemetry won't work till after there's a new release of INAV (probably v2.1) that includes the new Crossfire flight modes.  I may release a test version of Lua Telemetry that could work with Crossfire before INAV v2.1 is released, but that all depends.  Keep in mind that I don't have Crossfire so this is something that may not happen quickly, and I'm at the mercy of the INAV development team and the willingness of testers.

To accomplish this, I'll need help with someone running Crossfire and the HEX file in this issue.  You provided some old transmitter telemetry logs but those don't include the flight mode and were not created using the HEX file in this issue.  I'd need telemetry log files from while this HEX file is in loaded in INAV and with the flight mode sensor being recorded.

Maybe I'm not being clear what I need?  Or maybe not explaining what the HEX file is doing?

**Comment by teckel12** [comment]
> It doesnt show anything in the lua. There is no. Flight mode or anything

To me, this means you get a totally blank screen.  "doesnt show anything" means 100% blank screen.  Maybe a screen shot?  Even if you have nothing setup, it should show *something*, or an error, or something.

**Comment by KnuckleUpFPV** [comment]
Ok i guess i misunderstood what you were trying to say. I dont know what to look for because flight modes were already reported to crossfire. I loaded your omnibus hex and see no change in whats reported in flight modes when switches are changed. It shows the same thing. So before I updated to your hex the fm sensor still reports the same thing after changing to your hex.

**Comment by teckel12** [comment]
@KnuckleUpFPV Only 3 flight modes were reported before my changes:

```
ACRO
ANGL
HOR
```

After my changes, the flight modes outlined in the above posts will be shown. I need the telemetry logs to make Crossfire compatible with Lua Telemetry.

**Comment by KnuckleUpFPV** [comment]
If you mean those 3 modes in that box. They were already reported. If you are talking about something else im not seeing anything but those modes. As long as it doesnt rain tomorrow I should have some new logs for you. If you can build me a firmware for the matek f405 not the se I can run that as well. Unless just the omnibus telemetry is enough for you.

**Comment by KnuckleUpFPV** [comment]
Im using the hex in this thread. You will get your logs tomorrow as long as it doesnt rain. There is no option to only record part of the telemetry. Unless you go through deleting sensors pre-flight. Which would be an odd thing to do. I even hit discover sensors today to see if it added anything or found anything else. Keep your fingers crossed for no rain and I will fly on lunch.

**Comment by KnuckleUpFPV** [comment]
And its raining. Just my luck. Supposed to clear up by tomorrow. If there is a break in the rain during my lunch break, I will still put the wing up and go through the modes

**Comment by jorgermp** [comment]
> @jorgermp After INAV merges my changes, Lua Telemetry will work with Crossfire on the X9D, X7 and Horus transmitters.
> 
> The issue was never with Lua Telemetry, it was that Crossfire telemetry provided rudimentary flight information which doesn't give enough info for Lua Telemetry to function. Like actually knowing when the craft is armed as an example. This changes once INAV is updated in the next release.

thanks @teckel12 I had been using the inav screen in my qx7 with frsky R9 system, but now with crossfire don't work.

**Comment by jorgermp** [comment]
@KnuckleUpFPV thanks!!

**Comment by teckel12** [comment]
@KnuckleUpFPV The updated HEX doesn't add a new sensor, it just expands the existing `FM` (flight mode) sensor.  Previously, it would only show the following:

```
ACRO
ANGL
HOR
!FS
```

With the updated HEX (and INAV v2.1) the `FM` sensor should be expanded to show the following:

```
OK
WAIT
AIR
ACRO
ANGL
HOR
MANU
RTH
HOLD
3CRS
CRS
AH
WP
!FS!
!ERR
```

It's been a little challenging to nail down the telemetry sensor names for Crossfire as well.  I believe the following are returned:

```
FM
Sats
GPS
GSpd
Alt
Roll
Ptch
Yaw
1RSS
2RSS
RxBt
```

But, I also see `Curr` and `Capa` but want to be sure those are also Crossfire standards.  Also, the INAV Crossfire telemetry code exports a `Fuel` sensor value (battery remaining percentage) but I don't see that sensor in your logs.

Finally, the multiple RSSI and signal noise values need to be sorted out.  I could just show the highest of either `1RSS` or `2RSS`.  However, maybe showing `RSNR` (which I assume is receiver signal to noise) would be a better value to show?  Insight from Crossfire pilots on which signal value is the most useful would be helpful.

**Comment by KnuckleUpFPV** [comment]
We also have heading and one other sensor. I have to double check. For rssi you would want the rqly. Thats the link strength at the receiver. Where are you located? Once my replacement full size crossfire gets here, i can send you my micro crossfire for testing. Its been raining all day so couldnt get any flights in. Let me get my transmitter out and let me get the full sensor list for you. 

Ptch
Roll
Yaw
FM
1RSS
2RSS
RQly
RSNR
RFMD
TPWR
TRSS
TQly
TSNR
GPS
ALT
Sats
RxBt
Curr
Capa
GSpd
Hdg

Rxbt is battery voltage reported by fc
Rfmd is the mode crossfire is in. Rfmd2 is 150hz mode rfmd1 is 50hz mode, rfmd 0 is 4hz mode.
Tpwr is current transmit power of the crossfire. Currently tpwr shows 0 when set to 250mw on micro crossfire. All other power settings it shows appropriately.

The crossfire also tracks every switch and knob position. Weather is supposed to be clear tomorrow. So should be able to get a flight or two in. 

I do see wait reported under fm. I will be able to get more info for you tomorrow. 


We use the lq mapped to a channel for our rssi. You can choose rssi, rssi/lq,  or just lq. 

As far as fuel I only saw that with frsky setup. Which the crossfire can use sbus. I may try to use the sbus setting and see if the lua will show telemetry. I have a spare fc and rx to do some bench testing.

**Comment by teckel12** [comment]
@KnuckleUpFPV Thanks a lot for the info! I'll be really interested to see what the SBUS telemetry setting does (be sure to search for new sensors).

I'm the US (Toledo, Ohio). A loaner Crossfire would probably make this much easier. If that doesn't work out, I'll reach out to TBS and see if they'd be willing to donate hardware for development. I've had some success by just asking.

Doesn't the `RQly` sensor typically show 100%? It seemed that it wasn't very linear. Like it would show 100% and then drop to 40%. While RSSI would be much more granular. Maybe I'm mistaken or was looking at a poor example.

Thanks again.

**Comment by KnuckleUpFPV** [comment]
Yeah sorry, the rqly will hold at 99 until it drops into rfmd1 50hz mode. In rfmd2 it keeps the rqly at 99 because there is no fear of failsafing. Its not until you get way far out that it drops into rfmd1 and then link quality becomes important to know. Im from ohio. Just east of you. I live in Virginia right now. Once i get my replacement full size i can get a micro to you. Trappy would probably be willing to hook you up with a module. Anything that improves the usability of crossfire he loves.

**Comment by KnuckleUpFPV** [comment]
Yeah man the weather has been absolute crap here. Went from crazy high wind
to rain. Fingers crossed for tomorrow. The extra modes are definitely
showing from bench test.

**Comment by teckel12** [comment]
@KnuckleUpFPV Not great weather here either, warm today, but windy then rain. I don't expect much flying time over the winter in northern Ohio. Good time to build new quads and fix/upgrade others.

I feel the extra modes will be helpful to all Crossfire pilots that show the flight mode telemetry sensor. Not just for Lua Telemetry.

**Comment by KnuckleUpFPV** [comment]
I was able to get one flight in on lunch. Ill pull the sd log when I get home. I know its not much but its a start. I tried to go through all the modes. Everything except launch mode. It was to windy by the time it was my lunch break. Which is crazy because it was beautiful out earlier. Wind beat the snot out of my wing. Rth didnt engage first time I hit it. Made me panic a bit. Second time it went just fine.

**Comment by KnuckleUpFPV** [comment]
Sorry for the delay. Got tied up yesterday


[XFIRENANO-2018-12-03.zip](https://github.com/iNavFlight/LuaTelemetry/files/2646462/XFIRENANO-2018-12-03.zip)

**Comment by KnuckleUpFPV** [comment]
Looking at the log it doesnt seem to show the flight mode changing, even though the readout on the transmitter shows them.

**Comment by teckel12** [comment]
@KnuckleUpFPV I *thought* I posted a reply to this issue but I guess I didn't save it?

Anyway... I did look at your logs, and noted that the flight mode stays at `0`, probably because it's not provisioned for a string. I'm trying to remember what the rest of my message was... 

I'm getting my Crossfire running. This should allow a much speedier development. Still, the problem I'm having is that the OpenTX simulator doesn't receive unknown telemetry values. Most of the telemetry values Crossfire sends are not the same FrSky values. Therefore, there's no possible way to test Crossfire in the simulator.

This makes the development process painful. Write code, have the simulator compile the code, connect the transmitter, copy the code, disconnect transmitter, test with craft, repeat. This process is much more complex than S.Port, which is just write code, reset simulator, repeat. Oh well...

**Comment by KnuckleUpFPV** [comment]
Im sorry tim. I wish I could help you out. I wonder how that greer guy got his working. I wish he would port it to qx7 while you work on your masterpiece. I was fiddling with it with my buddys r9m. Its amazing how fast your lua responds to movement from the flight controller. You could literally fly by instrument.

**Comment by teckel12** [comment]
@KnuckleUpFPV It's just not an idea development environment.

The other script was just changing the sensor names and that's it. Lua Telemetry has a LOT more complexity and flight status monitoring to deal with. Also, it needs to work with S.Port, F.Port, D-series, and Crossfire on all Taranis and Horus transmitters. MUCH easier to make another script work just with Crossfire on just one transmitter. To have a single code base work with everything is vastly more complex, and the flight control built into Lua Telemetry make it an order of magnitude more complex.

Anyway, the main issue has always been not having Crossfire telemetry. People would ask for support, but that's nearly impossible to develop without having Crossfire. What's happening now is the preliminary changes required with INAV while I'm setting up my Crossfire build. We have time, as Crossfire support can't be released till INAV 2.1 anyway. I expect a Crossfire compatible version of Lua Telemetry soon after the INAV 2.1 release (if the parts I ordered from China arrive when expected).

**Comment by KnuckleUpFPV** [comment]
Well my replacement full size should be here within a week. So if my micro would help you sooner than you waiting on the slow boat, I can send it. I would send a receiver as well obviously.

**Comment by KnuckleUpFPV** [comment]
Wrap it in bubble wrap and toss it out the window. Lol.  On a more serious note. If you can build me a matekf405 target, I can test it on two wings. I wish there was a foxeer f405 target. Got a new fc and would love to put inav on it.

**Comment by teckel12** [comment]
@KnuckleUpFPV I'm also working on adding MSP to Crossfire which will allow VTx control from the transmitter. I can make you a HEX for the Matek405 that includes that too. Do you need it for the Matek405SE?

**Comment by KnuckleUpFPV** [comment]
Just the regular matekf405. Its an aio. The one with the sensitive gyro. Adding msp would be great. But not absolutely needed. We can change via osd or if using smart audio direct from the crossfire lua. The furious 2.4ghz vtx control should be implemented to inav soon.

**Comment by teckel12** [comment]
> Just the regular matekf405. Its an aio. The one with the sensitive gyro. Adding msp would be great. But not absolutely needed. We can change via osd or if using smart audio direct from the crossfire lua. The furious 2.4ghz vtx control should be implemented to inav soon.

@KnuckleUpFPV Here's the HEX for the Matek F405 with the Crossfire code:

[inav_2.1.0_MATEKF405.hex.zip](https://github.com/iNavFlight/LuaTelemetry/files/2665459/inav_2.1.0_MATEKF405.hex.zip)

**Comment by KnuckleUpFPV** [comment]
You are the man. If I could figure out how to do this I wouldn't need you to be the legend you are. I will get it loaded up asap.

**Comment by jorgermp** [comment]
@teckel12 can you make a HEX for matek f405 wing?, the target is MATEKF405SE for test Crossfire codo?

**Comment by teckel12** [comment]
@jorgermp Yup! Will make a build later this evening.

**Comment by teckel12** [comment]
@jorgermp Here ya go:

[inav_2.1.0_MATEKF405SE.hex.zip](https://github.com/iNavFlight/LuaTelemetry/files/2670367/inav_2.1.0_MATEKF405SE.hex.zip)

**Comment by teckel12** [comment]
Got Crossfire running on my new long-range build sending to the Horus transmitter. Using the latest Crossfire changes to INAV. Build still has a way to go before I can really call it a development environment. But I am getting Crossfire telemetry so I'm well on the way.

**Comment by ZinhoRacing** [comment]
What is your setup ? I m free to help if needed.

**Comment by KnuckleUpFPV** [comment]
This is fantastic news! Its been raining here non stop. Killing my drive to fly.

**Comment by KnuckleUpFPV** [comment]
If you are running 7 inch props get a titan 5200mah lithium ion. I run one
on my 6 inch and get 30 min of flying. You should get the same or longer.

**Comment by teckel12** [comment]
@KnuckleUpFPV I do have a 5200mAh I just use for a bench battery. I'm planning on testing how it flies with such a heavy battery. Glad to hear that you've had success. I'm sure acro flying is out with it.

**Comment by KnuckleUpFPV** [comment]
Believe it or not I can still do flips and rolls. My amp draw is too high
if I punch it though. If I had low kv and large props it would definitely
be doable. Im running 2205 2300kv chee pass motors from kwad box. They dont
make them anymore.

**Comment by teckel12** [comment]
Latest INAV build that adds Crossfire MSP support:

[inav_2.1.0_MATEKF405.hex.zip](https://github.com/iNavFlight/LuaTelemetry/files/2687986/inav_2.1.0_MATEKF405.hex.zip)
[inav_2.1.0_MATEKF405SE.hex.zip](https://github.com/iNavFlight/LuaTelemetry/files/2687987/inav_2.1.0_MATEKF405SE.hex.zip)

I'll be testing this on my Crossfire build this evening and posting more info so others could see what it does and what the capabilities are.

**Comment by KnuckleUpFPV** [comment]
Would you mind throwing up an omnibus pro firmware as well please?

**Comment by teckel12** [comment]
@KnuckleUpFPV

[inav_2.1.0_OMNIBUSF4PRO.hex.zip](https://github.com/iNavFlight/LuaTelemetry/files/2690679/inav_2.1.0_OMNIBUSF4PRO.hex.zip)

**Comment by teckel12** [comment]
I made some slight flight mode changes to the latest build.

1) `WAIT` doesn't show if extra arming safety is set to off (as `WAIT` is designed to indicate that it's waiting for a GPS lock and it should never wait for a GPS lock if extra arming safety is off)
1) New flight mode `AH` for altitude hold (altitude hold assumes ANGLE mode)
1) New flight mode `HRST` for home reset, which really isn't a flight mode, but indicates the setting of a new home position (this is a temporary flight mode, just while the home reset trigger is set)

The HEX files can be downloaded from the following message:

https://github.com/iNavFlight/inav/issues/3886#issuecomment-448796741

**Comment by KnuckleUpFPV** [comment]
Will home reset be effected by the fact that I use first arming? Ive accidentally disarmed in the air before like an idiot. So now everything is set to first arm.

**Comment by teckel12** [comment]
@KnuckleUpFPV I'm not changing the home reset logic, just displaying when home reset is triggered. This would only be displayed when the user initiates a home reset (like via a switch). It sounds like you don't have a home reset switch, so you would never get a home reset flight mode notification.

**Comment by KnuckleUpFPV** [comment]
Ok im tracking now. That makes sense. Yeah no home reset. I let it set first arm as home position. So if I accidentally disarm it knows where home is.

**Comment by teckel12** [comment]
@KnuckleUpFPV Yeah, has nothing to do with your setup. Only for those with a home reset switch. I also only set home at first arm, but have a home reset switch because sometimes I like to land in another location.

**Comment by Giblets-FPV** [comment]
While I'm obviously new to this I would be more than happy to help. I'll start recording logs for you and getting them uploaded. Since my x-lite can only run Crossfire (internal crossfire mod), any models and logs uploaded will be on crossfire. But I'll need a hex for the matek f722-std to be of any help.

**Comment by teckel12** [comment]
@Giblets-FPV I'm working on getting the Horus release out, then I'll focus on Crossfire. I now have a working Crossfire setup, so logs are no longer needed.

Probably in the next week or two I'll have a Crossfire build of Lua Telemetry. At that time, I'll also post INAV firmware that supports the Crossfire enhancements (including MSP). I'll include the Matek F722-STD hex.

**Comment by teckel12** [comment]
@Giblets-FPV Here's an INAV build with the Crossfire changes:

[inav_2.1.0_MATEKF722.hex.zip](https://github.com/iNavFlight/LuaTelemetry/files/2729954/inav_2.1.0_MATEKF722.hex.zip)

**Comment by teckel12** [comment]
@ZinhoRacing @KnuckleUpFPV @Giblets-FPV The development build of Lua Telemetry is now working with Crossfire! It's not 100%, but mostly functional. Instructions for installing the development build:

https://github.com/iNavFlight/LuaTelemetry/wiki/Upgrade-to-Development-Build

### Known issues

* Heading isn't working correctly, or at least not for me. Really strange results. Almost seems Crossfire isn't providing INAV heading but is doing its own thing (maybe based on GPS)? 
* The HDOP is wrong, ignore it (it will be removed/replaced as Crossfire doesn't provide GPS quality info)
* GPS altitude always shows zero. That will also be removed as Crossfire doesn't provide this either

### Possible issues

* Distance to home seems to work, but it's being calculated/estimated based on launch vs current GPS coordinates
* I wouldn't totally trust the map because the heading isn't working
* I only tested it on Horus so Taranis is totally untested (but I'd love to get your feedback)!

Enjoy and please provide feedback!

**Comment by ZinhoRacing** [comment]
Great @teckel12 

Heading is Yaw value (you need to convert Yaw value that is radian to degre), you can check it on Inav configurator, page Setup, heading value.

I can't make it work, it stays with in no telem state.
I followed the wiki. is there any setup to activate CRSF ?

**Comment by teckel12** [comment]
@ZinhoRacing Then what's `Hdg`? That's the standard heading telemetry value for OpenTX, and is already in degrees.  I can change it to `Yaw`, but still, what's `Hdg`? 

Crossfire selection is automatic. It looks that you **don't** have an `RSSI` sensor and that you have an `FM` sensor. Maybe this was a little short-sighted, as it assumes default sensor names. Maybe it should only look for the `FM` sensor as that's required to work. Making it an option that needs to be selected doesn't seem necessary, as that would use more memory and require an extra step.

**Comment by ZinhoRacing** [comment]
Hdg should be the GPS ground course (so heading), you are right. But it doesn't work, I tryed to find why (in openTX code and Inav code) but I couldn't find the problem ...
Inav code : 
```
/*
0x02 GPS
Payload:
int32_t     Latitude ( degree / 10`000`000 )
int32_t     Longitude (degree / 10`000`000 )
uint16_t    Groundspeed ( km/h / 10 )
uint16_t    GPS heading ( degree / 100 )
uint16      Altitude ( meter ­1000m offset )
uint8_t     Satellites in use ( counter )
*/
void crsfFrameGps(sbuf_t *dst)
{
    // use sbufWrite since CRC does not include frame length
    sbufWriteU8(dst, CRSF_FRAME_GPS_PAYLOAD_SIZE + CRSF_FRAME_LENGTH_TYPE_CRC);
    crsfSerialize8(dst, CRSF_FRAMETYPE_GPS);
    crsfSerialize32(dst, gpsSol.llh.lat); // CRSF and betaflight use same units for degrees
    crsfSerialize32(dst, gpsSol.llh.lon);
    crsfSerialize16(dst, (gpsSol.groundSpeed * 36 + 50) / 100); // gpsSol.groundSpeed is in cm/s    
	crsfSerialize16(dst, DECIDEGREES_TO_CENTIDEGREES(gpsSol.groundCourse)); // gpsSol.groundCourse is 0.1 degrees, need 0.01 deg
    const uint16_t altitude = (getEstimatedActualPosition(Z) / 100) + 1000;
    crsfSerialize16(dst, altitude);
    crsfSerialize8(dst, gpsSol.numSat);
}
```
So you also see that the Altitude reported is the actual Altitude (not GPS altitude see level). open TX labelled it Alt so not so bad. (inav and OpenTx are using CRSF GPS Alt for Alt sensor)

I'm using default OpenTx sensor names. So should work ... I'm trying to find where is the problem.

**Comment by ZinhoRacing** [comment]
Ok, find it !!!
In french, OpenTX label for FM is PV (for Phase de Vol, french translation !!! ) this is the only language that is not using FM for Flight Mode ...

Unlucky .

![image](https://user-images.githubusercontent.com/41761785/50730734-78048d00-1154-11e9-8690-1ff21d6709f2.png)

**Comment by teckel12** [comment]
@ZinhoRacing That's hilarious. I assume that's the Crossfire source?

There's an easy solution on my end. But for the time-being you can just change the sensor name to `FM` in OpenTX.

**Comment by teckel12** [comment]
Also... If there's any issues or suggestions with the development branch and Crossfire support, if you could open a new issue that would be appreciated. This issue is **VERY** long at this point (124 replies) :dizzy_face:

**Comment by teckel12** [comment]
@advancetom Crossfire support added, see above for instructions.

**Comment by KnuckleUpFPV** [comment]
Sounds great man. I will reinstall the lua today. Hoping the wind dies down some. Might be able to get out and test it.

**Comment by KnuckleUpFPV** [comment]
Lua works awesome, except for the alarms im getting constantly. I cant figure out which one is causing it. Its very fast on the qx7. Can definitely fly from instruments.

**Comment by TwoToneEddy** [comment]
Hi, just so I'm clear, has this only been implemented on the development branch at the moment? Will it be released soon or should I just build from development?

**Comment by teckel12** [comment]
@TwoToneEddy It's only in the development branch of Lua Telemetry.  The reason is that INAV 2.1.0 isn't released yet, which is required for Crossfire compatibility with Lua Telemetry.  I don't want to release a version of Lua Telemetry that includes features that don't work with stable flight controller firmware.

But, using INAV 2.1.0 RC3 and the development branch of Lua Telemetry will get you Crossfire support.

It's super simple to upgrade to the development branch, and there's a Wiki on it:

https://github.com/iNavFlight/LuaTelemetry/wiki/Upgrade-to-Development-Build

**Comment by KnuckleUpFPV** [comment]
Tim don't know if you saw but gps speed and heading were added to crossfire telemetry for opentx

**Comment by teckel12** [comment]
@KnuckleUpFPV This must have been added a while ago because I've had GPS speed and heading since I've had Crossfire.  Although, the heading is only correct with the `Yaw` sensor.  The `Hdg` sensor doesn't seem to work all the time.  Also, the `Yaw` value is incorrect as it overflows.  I've reported this to TBS but I'm not sure if they took it seriously.  I had to write special code to compensate for the 16 bit integer rollover that's happening on `Yaw`.

**Comment by teckel12** [comment]
@KnuckleUpFPV Was something just released that referenced these sensors?

**Comment by teckel12** [comment]
Just verified, these sensors were added back in v2.31 on 2018-04-18.  As I mentioned, however, `Hdg` isn't correct so I'm using `Yaw` and `Yaw` has an integer overflow bug which TBS has not yet corrected (as of v2.88) but Lua Telemetry has compensated for.

This is the code that detects if there's an integer overflow and adjusts the heading:

```
	-- The following is done due to an int rollover bug in the Crossfire protocol
	local tmp = getValue(data.hdg_id)
	if tmp < -0.27 then
		tmp = tmp + 0.27
	end
```
